### PR TITLE
examples: additions and fixes

### PR DESCRIPTION
**Pull request by petersalomonsen**
here's the example fixes / additions from my emscripten PR as requested in https://github.com/libgit2/libgit2/pull/5345#issuecomment-589931956

- add example for git commit
- fix example for git add
- add example for git push

### CODE REVIEW

**Comment by pks-t** [comment]
I will review this after v1.0 is released. Please feel free to ping me after the release if I forget.

File: examples/commit.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * - Example of performing a git commit with a comment
+ * 
+ */
+int lg2_commit(git_repository *repo, int argc, char **argv)
+{
+	const char *opt = argv[1];
+	const char *comment = argv[2];
+	int error;
+
+    git_oid commit_oid,tree_oid;
```
Comment (pks-t) [inline examples/commit.c:37]: This line is indented with spaces

File: examples/commit.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	/* Validate args */
+	if (argc < 3 || strcmp(opt, "-m")!=0) {
+		printf ("USAGE: %s -m <comment>\n", argv[0]);
+		return -1;
+	}
+
+	git_revparse_ext(&parent, &ref, repo, "HEAD");
+	git_repository_index(&index, repo);	
+	git_index_write_tree(&tree_oid, index);
+	git_index_write(index);
```
Comment (pks-t) [inline examples/commit.c:55]: All of the above commands should be checked for errors.

File: examples/commit.c [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	git_tree *tree;
+	git_index *index;	
+	git_object *parent = NULL;
+	git_reference *ref = NULL;
+	git_signature *signature;	
+
+	(void)repo; /* unused */
+
+	/* Validate args */
+	if (argc < 3 || strcmp(opt, "-m")!=0) {
```
Comment (pks-t) [inline examples/commit.c:47]: Please add a space before and after `!=`

File: examples/commit.c [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		NULL,
+		comment,
+		tree,
+		parent ? 1 : 0, parent);
+		
+	if (error != 0) {
+		const git_error *err = git_error_last();
+		if (err) printf("ERROR %d: %s\n", err->klass, err->message);
+		else printf("ERROR %d: no detailed info\n", error);
+	}
```
Comment (pks-t) [inline examples/commit.c:82]: You can use `check_lg2` in general for any kind of error checking.

File: examples/push.c [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+int lg2_push(git_repository *repo, int argc, char **argv) {
+    // get the remote.
+	int error;
+	git_push_options options;
+	git_remote* remote = NULL;
+	char *refspec = "refs/heads/master";
+    const git_strarray refspecs = {
+		&refspec,
+		1,
+	};
```
Comment (pks-t) [inline examples/push.c:41]: There's some formatting errors above with regards to whitespace. We also always use `/* */` even for single-line comments, same below.

File: examples/push.c [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+	// do the push
+
+	error = git_remote_push( remote, &refspecs, &options);
+
+	if (error != 0) {
+		const git_error *err = git_error_last();
+		if (err) printf("ERROR %d: %s\n", err->klass, err->message);
+		else printf("ERROR %d: no detailed info\n", error);
+	}
```
Comment (pks-t) [inline examples/push.c:62]: If you use `check_lg2`, then you don't need this block. so please also use it fo r `git_remote_lookup` and `git_push_options_init`.

**Comment by petersalomonsen** [comment]
Thanks for reviewing @pks-t . I believe the issues are resolved now.  For `commit` I had to handle `GIT_ENOTFOUND` differently ( empty repositories have no commits ), but elsewhere I've used `check_lg2` for error handling. Please let me know if there are still issues.

File: examples/commit.c [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+	int error;
+
+	git_oid commit_oid,tree_oid;
+	git_tree *tree;
+	git_index *index;	
+	git_object *parent = NULL;
+	git_reference *ref = NULL;
+	git_signature *signature;	
+
+	(void)repo; /* unused */
```
Comment (pks-t) [inline examples/commit.c:44]: This line can be removed, you use the repo ;)

File: examples/commit.c [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+	/* Validate args */
+	if (argc < 3 || strcmp(opt, "-m") != 0) {
+		printf ("USAGE: %s -m <comment>\n", argv[0]);
+		return -1;
+	}
+
+	error = git_revparse_ext(&parent, &ref, repo, "HEAD");
+	if ( error == GIT_ENOTFOUND) {
```
Comment (pks-t) [inline examples/commit.c:53]: There's a stray space here in between `( error`

File: examples/commit.c [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+	check_lg2(git_repository_index(&index, repo), "Could not open repository index", NULL);
+	check_lg2(git_index_write_tree(&tree_oid, index), "Could not write tree", NULL);;
+	check_lg2(git_index_write(index), "Could not write index", NULL);;
+
+	git_index_free(index);
+
+	check_lg2(git_tree_lookup(&tree, repo, &tree_oid), "Error looking up tree", NULL);
+	
+	git_signature_default(&signature, repo);
```
Comment (pks-t) [inline examples/commit.c:70]: This also requires error checking.

File: examples/push.c [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+	if (argc > 1) {
+		printf ("USAGE: %s\n\nsorry, no arguments supported yet\n", argv[0]);
+		return -1;
+	}
+
+	check_lg2(git_remote_lookup( &remote, repo, "origin" ), "Unable to lookup remote", NULL);
+	
+	check_lg2(git_push_options_init( &options, GIT_PUSH_OPTIONS_VERSION ), "Error initializing push", NULL);
+
+	check_lg2(git_remote_push( remote, &refspecs, &options), "Error pushing", NULL);
```
Comment (pks-t) [inline examples/push.c:52]: There's a few more stray spaces in these commands :)

File: examples/commit.c [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+		const git_error *err = git_error_last();
+		if (err) printf("ERROR %d: %s\n", err->klass, err->message);
+		else printf("ERROR %d: no detailed info\n", error);
+	}
+
+	check_lg2(git_repository_index(&index, repo), "Could not open repository index", NULL);
+	check_lg2(git_index_write_tree(&tree_oid, index), "Could not write tree", NULL);;
+	check_lg2(git_index_write(index), "Could not write index", NULL);;
+
+	git_index_free(index);
```
Comment (pks-t) [inline examples/commit.c:66]: I'd move this line before the `return error` to not disturb the program's business logic flowing and make it easier to understand

**Comment by petersalomonsen** [comment]
yeah, no problem being nitpicky :-) quality is important, and I feel a bit novice on the code conventions here - so very happy for your guidance :-) Think I got the issues resolved now, but let me know if there are more :-)

**Comment by pks-t** [comment]
> yeah, no problem being nitpicky :-) quality is important, and I feel a bit novice on the code conventions here - so very happy for your guidance :-) Think I got the issues resolved now, but let me know if there are more :-)

That's on us, though. We don't have a style guide anywhere, so newcomers are left guessing. We should definitely have one, but I've been saying that for a few years now and never just sat down and wrote one.

**Comment by aaronfranke** [comment]
The push example always fails for me. I always get `-1` returned from `git_remote_push`. I tested pushing to remotes on GitHub, GitLab, and a local repo on my desktop. I tried the refspec as `refs/heads/master` and `refs/heads/master:refs/remotes/myremote/master`, both fail. The `git_remote_lookup` and `git_push_options_init` methods always succeed and return 0.

**Pull request merged**